### PR TITLE
Thisinq

### LINKED ISSUE #128: cannot have $where in $elemMatch
Can't canonicalize query: BadValue $elemMatch cannot contain $where expression ns:testdb.operationalProductRelationship query:{ skuTree: { $elemMatch: { $and: [ { $where: function() {if(this.parent.skuCode == this.$parent.rootProduct.skuCode... }, { $where: function() {if(this.parent.inventoryOperatingUnitCode == this.$parent.... } ] } } }
[mongod output] com.redhat.lightblue.client.http.LightblueHttpClientException: Error sending lightblue request: POST /find/operationalProductRelationship/1.0.3-SNAPSHOT, body: {"query":{"array":"skuTree","elemMatch":{"$and":[{"field":"parent.skuCode","op":"$eq","rfield":"$parent.rootProduct.skuCode"},{"field":"parent.inventoryOperatingUnitCode","op":"$eq","rfield":"$parent.rootProduct.inventoryOperatingUnitCode"}]}},"projection":[{"field":"*","include":true,"recursive":true}]}

### LINKED ISSUE #333: elemMatch fails with field $in array
Request:
```
{
    "objectType": "myEntity",
    "projection": [
        {
            "field": "*",
            "include": true,
            "recursive": true
        }
    ],
    "query": {
    	"array": "my-array",
    	"elemMatch": {
    		"field": "$this",
    		"op": "$in",
    		"rfield": "$parent.my-array"
    	}
    }
}

```

Response:
```
{
  "status": "COMPLETE",
  "modifiedCount": 0,
  "matchCount": 0,
  "processed": [],
  "errors": [
    {
      "objectType": "error",
      "context": "rest/FindCommand/myEntity/find(myEntity)/find/translateQuery",
      "errorCode": "mongo-translation:invalid-field",
      "msg": "{\"array\":\"my-array\",\"elemMatch\":{\"field\":\"$this\",\"op\":\"$in\",\"rfield\":\"$parent.my-array\"}}"
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by bserdar**
Fix #333: Write javascript query for this case
Fix #128: If we need a $where under array elemMatch, write the whole thing in Javascript

### CODE REVIEW

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/js/JSQueryTranslator.java [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+     *   var r1=false;
+     *   r1=r0.test(field);
+     *   return r1;
+     * </pre>
+     */
+    private Block translateRegexMatchExpression(Context ctx,RegexMatchExpression query) {
+        FieldTreeNode fieldMd=ctx.contextNode.resolve(query.getField());
+        if(fieldMd.getType() instanceof StringType) {
+            Name fieldName=ctx.varName(new Name(query.getField()));
+            String regexVar=ctx.topLevel.newGlobal(ctx,String.format("new RegExp(\"%s\",\"%s\")",
```
Comment (paterczm) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/js/JSQueryTranslator.java]: Those \" are a typo?

**Comment by bserdar** [comment]
Are you asking about "new RegExp(\"%s\",\"%s\")"? It should write:

var v=new RegExp("pattern","options")

**Comment by paterczm** [comment]
> Are you asking about "new RegExp(\"%s\",\"%s\")"?

I got confused and removed my initial comment. The problem is that RegExp does not expect the "/regex/" notation, but this is what it gets and matching does not work.

Other than that, the changes look good to me. Created a pr334-fix branch with a fix for the above and some additional unit tests.

**Comment by paterczm** [comment]
My bad, I defined the regex in the query incorrectly. All looks well. Merging.

**Pull request merged**